synapse/storage/databases/main/state.py

```diff
@@ -133,6 +145,36 @@ def get_room_version_id_txn(self, txn: LoggingTransaction, room_id: str) -> str:
 
         return room_version
 
+    async def get_metadata_for_events(
+        self, event_ids: Collection[str]
+    ) -> Dict[str, EventMetadata]:
+        """Get some metadata (room_id, type, state_key) for the given events."""
+
+        clause, args = make_in_list_sql_clause(
+            self.database_engine, "e.event_id", event_ids
+        )
+
+        sql = f"""
+            SELECT e.event_id, e.room_id, e.type, e.state_key FROM events AS e
+            LEFT JOIN state_events USING (event_id)
+            WHERE {clause}
```

Comment:
hrmph. This is exactly the sort of thing https://github.com/matrix-org/synapse/issues/11496 is supposed to fix, by making it so we don't have to hit two tables.

I suppose we're going to have to have fallback code anyway, so this is fine for now, but I'd really love it if we could find some tuits to keep progressing that.